Slack integration - first draft of blog post (#2938)
pulumi/pulumi-hugo: themes/default/content/blog/slack-integration/index.md

| `update_succeeded`     | `stack_update`  | Both                       | When a stack `update` succeeds.   |
| `update_failed`        | `stack_update`  | Both                       | When a stack `update` fails.      |
| `destroy_succeeded`    | `stack_update`  | Both                       | When a stack `destroy` succeeds.  |
| `destroy_failed`       | `stack_update`  | Both                       | When a stack `destroy` fails.     |
| `refresh_succeeded`    | `stack_update`  | Both                       | When a stack `refresh` succeeds.  |
| `refresh failed`       | `stack_update`  | Both                       | When a stack `refresh` fails.     |
| `deployment_queued`    | `deployment`    | Both                       | When a deployment is queued.      |
| `deployment_started`   | `deployment`    | Both                       | When a deployment starts running. |
| `deployment_succeeded` | `deployment`    | Both                       | When a deployment succeeds.       |
| `deployment_failed`    | `deployment`    | Both                       | When a deployment fails.          |

## Wrapping up

> We hope you and your team can streamline deployment related notifications to where you spend your time. Our webhooks improvements enable ChatOps workflows and more visibility into your infrastructure. As always, please let us know if you have feedback on the feature by opening an issue in the Pulumi Cloud requests repository or if you have other features you would like to see in Pulumi Cloud.

Author: ```suggestion
We hope you and your team can streamline deployment related notifications to where you spend your time. Our webhooks improvements enable ChatOps workflows and more visibility into your infrastructure. As always, please let us know if you have feedback on the feature by opening an issue in the [Pulumi Cloud requests repository](https://github.com/pulumi/pulumi-cloud-requests/issues/new/choose) or if you have other features you would like to see in Pulumi Cloud.
```

---

> title: "Announcing Slack and Deployment Notifications in Pulumi Cloud"

Author: ```suggestion
title: "Announcing Slack and Deployment Notifications for Pulumi Cloud"
```

---
title: "Announcing Slack and Deployment Notifications in Pulumi Cloud"

allow_long_title: true
# The date represents the post's publish date,
# and by default corresponds with the date this file was generated.
# Posts with future dates are visible in development,
# but excluded from production builds.
# Use the time and timezone-offset portions of of this value
# to schedule posts for publishing later.

> date: 2023-05-26T11:07:30-07:00

Reviewer: reminder to update this

# Use the meta_desc property to provide a brief summary
# (one or two sentences) of the content of the post,
# which is useful for targeting search results or social-media previews.
# This field is required or the build will fail the linter test.
# Max length is 160 characters.
meta_desc: Pulumi Cloud now has an easy to set up Slack integration, Pulumi Deployments notifications and event filtering.

# The meta_image appears in social-media previews and on the blog home page.
# A placeholder image representing the recommended format, dimensions and aspect ratio
# has been provided for you.
meta_image: meta.png

# At least one author is required.
# The values in this list correspond with the `id` properties
# of the team member files at /data/team/team.
# Create a file for yourself if you don't already have one.
authors:
    - meagan-cojocar
    - komal-ali

# At least one tag is required.
# Lowercase, hyphen-delimited is recommended.
tags:
    - features
    - pulumi-cloud

# See the blogging docs at https://github.com/pulumi/pulumi-hugo/blob/master/BLOGGING.md.
# for additional details,
# and please remove these comments before submitting for review.
---

> Today, we are excited to introduce a set of improvements to [Pulumi Cloud Webhooks](/docs/pulumi-cloud/webhooks), designed to streamline your deployment notifications to where you already spend your time, enabling faster response times to critical issues. We are also announcing two new improvements to our webhooks feature: Pulumi Deployments events and fine-grained filtering.

Reviewer 2: ```suggestion
Today, we are excited to introduce a set of improvements to [Pulumi Cloud Webhooks](/docs/pulumi-cloud/webhooks), designed to streamline your deployment notifications to where you already spend your time, enabling faster response times to critical issues. We are also announcing two new improvements to our webhooks feature: Pulumi Deployments events and fine-grained event filtering.
```

Reviewer 2: This sentence should also mention Slack specifically, I can't quite figure out the wordsmithing to get it in there.

<!--more-->

Teams using Pulumi Cloud have been setting up Slack notifications for their stacks using Pulumi Cloud Webhooks. Webhooks can attach to a Pulumi organization or a Pulumi stack. Starting today, customers can set up a Slack integration, for organization or stack notifications, with less steps and without needing to host the infrastructure themselves. We have also expanded the types of event notifications and the filtering options on both Pulumi Cloud Webhooks and for the Slack integration.

Pulumi Cloud Webhooks, including the Slack integration, are available to all Pulumi Cloud organizations.

## Slack integration

> ### Setiing up the Slack Integration

Reviewer 2: ```suggestion
### Setting up the Slack Integration
```

import java.nio.file.Paths;

public class App {
    public static void main(String[] args) {
        Pulumi.run(App::stack);
    }

    public static void stack(Context ctx) {
        var orgWebhook = new Webhook("orgWebhook", WebhookArgs.builder()
            .active(true)
            .displayName("#my-org-activity-channel")
            .organizationName("my-org")
            .payloadUrl("https://hooks.slack.com/services/...")
            .format("slack")
            .filters(
                "stack_created",
                "stack_deleted",
                "deployment_succeeded",
                "deployment_failed")
            .build());

        var stackWebhook = new Webhook("stackWebhook", WebhookArgs.builder()
            .active(true)
            .displayName("#my-stack-activity-channel")
            .organizationName("my-org")
            .projectName("webhook-examples")
            .stackName("dev")
            .payloadUrl("https://hooks.slack.com/services/...")
            .format("slack")
            .build());

        ctx.export("orgWebhookName", orgWebhook.name());
        ctx.export("stackWebhookName", stackWebhook.name());
    }
}
```

{{% /chooser %}}

> ## Pulumi Deployment Notifictation Events

Reviewer 2: ```suggestion
## Pulumi Deployment Notification Events
```

> ```

Reviewer 3: Missing a closing `choosable` for java, which is messing up the formatting for the rest of the post.

Reviewer 2: fixed this!

Pulumi Deployments is infrastructure deployments as a managed service. With Pulumi Deployments, you can run a Pulumi infrastructure as code action (a preview, update, destroy or refresh) inside Pulumi’s managed service. Pulumi provides scalability, observability and security for deployments. Both Slack-formatted and generic JSON webhooks in Pulumi Cloud now send notifications events on Pulumi Deployments statuses: when a deployment is queued, started, succeeds and fails.

> ## Filtering Notifictation Events

Reviewer 2: ```suggestion
## Filtering Notification Events
```

> Pulumi Deployments is infrastructure deployments as a managed service. With Pulumi Deployments, you can run a Pulumi infrastructure as code action (a preview, update, destroy or refresh) inside Pulumi’s managed service. Pulumi provides scalability, observability and security for deployments. Both Slack-formatted and generic JSON webhooks in Pulumi Cloud now send notifications events on Pulumi Deployments statuses: when a deployment is queued, started, succeeds and fails.

Reviewer 3: I would just link the first sentence from the deploy docs:

 Pulumi Deployments is a fully managed platform for managing infrastructure at scale.

Also, might just be useful to show a screenshot here of a deploy-specific slack notification just to break up the text. Maybe a queued notif?

---

> title: "Announcing Slack and Deployment Notifications in Pulumi Cloud"

Reviewer 4: The meta image has slightly different text. 
> ... for Pulumi Cloud

allow_long_title: true
# The date represents the post's publish date,
# and by default corresponds with the date this file was generated.
# Posts with future dates are visible in development,
# but excluded from production builds.
# Use the time and timezone-offset portions of of this value
# to schedule posts for publishing later.

> date: 2023-05-26T11:07:30-07:00

Reviewer 5: ```suggestion
date: 2023-05-30
```

# Use the meta_desc property to provide a brief summary
# (one or two sentences) of the content of the post,
# which is useful for targeting search results or social-media previews.
# This field is required or the build will fail the linter test.
# Max length is 160 characters.

> meta_desc: Pulumi Cloud now has an easy to set up Slack integration, Pulumi Deployments notifications and event filtering.

Reviewer 5: ```suggestion
meta_desc: Pulumi Cloud now has an easy way to set up Slack integration, Pulumi Deployments notifications and event filtering.
```

<!--more-->

Teams using Pulumi Cloud have been setting up Slack notifications for their stacks using Pulumi Cloud Webhooks. Webhooks can attach to a Pulumi organization or a Pulumi stack. Starting today, customers can set up a Slack integration, for organization or stack notifications, with fewer steps and without needing to host the infrastructure themselves. We have also expanded the types of event notifications and the filtering options on both Pulumi Cloud Webhooks and for the Slack integration.

Pulumi Cloud Webhooks, including the Slack integration, are available to all Pulumi Cloud organizations.

## Slack integration

### Setting up the Slack Integration

![Example of the notifications in Slack](slack.png)

> Before today Pulumi customers used Pulumi Cloud Webhooks to set up generic JSON webhooks. When events occur, we send a HTTP POST request to any registered listeners. Webhooks can then be used to send notifications to an app (like Slack), start running automated tests, or even to update another stack! We have now built support for Slack-formatted webhooks, which allow you to quickly enable notifications about your Pulumi stacks and organizations into your Slack workspace by simply providing a [Slack incoming webhook URL](https://api.slack.com/messaging/webhooks).

Reviewer 4: ```suggestion
Before today Pulumi customers used Pulumi Cloud Webhooks to set up generic JSON webhooks. When events occur, we send an HTTP POST request to any registered listeners. Webhooks can then be used to send notifications to an app (like Slack), start running automated tests, or even to update another stack! We have now built support for Slack-formatted webhooks, which allow you to quickly enable notifications about your Pulumi stacks and organizations into your Slack workspace by simply providing a [Slack incoming webhook URL](https://api.slack.com/messaging/webhooks).
```

You can either create your own Slack app (or use an existing one you may already have installed in your workspace) or you can use the link below to quickly get started with a predefined Slack app manifest.

> [Create a Slack app from manifest](https://api.slack.com/apps?new_app=1&manifest_yaml=display_information%3A%0A%20%20name%3A%20pulumi-slack-notifications%0A%20%20description%3A%20Funnel%20Pulumi%20webhooks%20to%20Slack%0A%20%20background_color%3A%20%22%238a3391%22%0Afeatures%3A%0A%20%20bot_user%3A%0A%20%20%20%20display_name%3A%20pulumi-slack-notifications%0A%20%20%20%20always_online%3A%20false%0Aoauth_config%3A%0A%20%20scopes%3A%0A%20%20%20%20bot%3A%0A%20%20%20%20%20%20-%20incoming-webhook%0Asettings%3A%0A%20%20org_deploy_enabled%3A%20false%0A%20%20socket_mode_enabled%3A%20false%0A%20%20token_rotation_enabled%3A%20false)

Reviewer 5: Ah! I just remembered you asked about this. Lemme see if I can get you a shortcode for it real quick. Mind if I push to this PR branch @meagancojocar?

![Setting it up in the UI](webhooks_2.gif)

There are Pulumi customers with hundreds to thousands of stacks, making setting up webhooks for each via the UI a time consuming task. In order to enable Pulumi notifications at scale, we have added Pulumi Cloud REST API endpoints for creating webhooks as well as [Pulumi Service Provider](https://www.pulumi.com/registry/packages/pulumiservice) support.

### Pulumi Cloud REST API

Create a webhook via the Pulumi Cloud REST API formatted for Slack, as shown in the example below.

```bash

curl \
  -H "Accept: application/vnd.pulumi+8" \
  -H "Content-Type: application/json" \
  -H "Authorization: token $PULUMI_ACCESS_TOKEN" \
  --request POST \
  --data '{
      "organizationName":"{organization}",
      "projectName":"{project}",
      "stackName":"{stack}",
      "displayName":"#some-slack-channel",
      "payloadUrl":"https://hooks.slack.com/services/...",
      "format": "slack",
      "active":true
  }' \
  https://api.pulumi.com/api/orgs/{organization}/{project}/{stack}/hooks

> 
> 

Reviewer 5: ```suggestion
```

```

### Pulumi Service Provider

The Pulumi Service provider allows you to create Pulumi Cloud resources via Pulumi. You can provision and manage webhooks, including Slack-formatted webhooks, using this provider.

> {{< chooser language "typescript,python,go,csharp,java,yaml" >}}

Reviewer 5: Using a trailing slash here will ensure the chooser uses (and sets) the user's language preferred language (otherwise it only affects the choosable items that it contains): 

```suggestion
{{< chooser language "typescript,python,go,csharp,java,yaml" />}}
```

{{% choosable language typescript %}}

```ts

> 

Reviewer 5: ```suggestion
```

import * as pulumi from "@pulumi/pulumi";
import { Webhook, WebhookFormat, WebhookFilters } from "@pulumi/pulumiservice";

const orgName = "my-org";

const orgWebhook = new Webhook("org-webhook", {

>   active: true,
>   displayName: "#my-org-activity-channel",
>   organizationName: orgName,
>   payloadUrl: "https://hooks.slack.com/services/...",
>   format: WebhookFormat.Slack,
>   filters: [
>       WebhookFilters.StackCreated,
>       WebhookFilters.StackDeleted,
>       WebhookFilters.DeploymentSucceeded,
>       WebhookFilters.DeploymentFailed,
>   ],
> });
> 
> const stackWebhook = new Webhook("stack-webhook", {
>   active: true,
>   displayName: "#my-stack-activity-channel",
>   organizationName: orgName,
>   projectName: pulumi.getProject(),
>   stackName: pulumi.getStack(),
>   payloadUrl: "https://hooks.slack.com/services/...",
>   format: WebhookFormat.Slack,
> })

Reviewer 5: ```suggestion
    active: true,
    displayName: "#my-org-activity-channel",
    organizationName: orgName,
    payloadUrl: "https://hooks.slack.com/services/...",
    format: WebhookFormat.Slack,
    filters: [
        WebhookFilters.StackCreated,
        WebhookFilters.StackDeleted,
        WebhookFilters.DeploymentSucceeded,
        WebhookFilters.DeploymentFailed,
    ],
});

const stackWebhook = new Webhook("stack-webhook", {
    active: true,
    displayName: "#my-stack-activity-channel",
    organizationName: orgName,
    projectName: pulumi.getProject(),
    stackName: pulumi.getStack(),
    payloadUrl: "https://hooks.slack.com/services/...",
    format: WebhookFormat.Slack,
});
```

export const orgWebhookName = orgWebhook.name;
export const stackWebhookName = stackWebhook.name;

> 

Reviewer 5: ```suggestion
```

```

{{% /choosable %}}

{{% choosable language yaml %}}

```yaml

name: webhook-examples
runtime: yaml
description: An example of provisioning a Pulumi Cloud Webhook via Pulumi Service Provider

resources:
  orgWebhook:
    type: pulumiservice:index:Webhook
    properties:
      active: true
      displayName: "#my-org-activity-channel"
      organizationName: my-org
      payloadUrl: "https://hooks.slack.com/services/..."
      format: slack
      filters:
        - stack_created
        - stack_deleted
        - deployment_succeeded
        - deployment_failed
  stackWebhook:
    type: pulumiservice:index:Webhook
    properties:
      active: true
      displayName: "#my-stack-activity-channel"
      organizationName: my-org
      projectName: webhook-examples
      stackName: dev
      payloadUrl: "https://hooks.slack.com/services/..."
      format: slack

outputs:
  # export the name of the webhooks
  orgWebhookName: ${orgWebhook.name}
  stackWebhookName: ${stackWebhook.name}

> 

Reviewer 5: ```suggestion
```

| `update_succeeded`     | `stack_update`  | Both                       | When a stack `update` succeeds.   |
| `update_failed`        | `stack_update`  | Both                       | When a stack `update` fails.      |
| `destroy_succeeded`    | `stack_update`  | Both                       | When a stack `destroy` succeeds.  |
| `destroy_failed`       | `stack_update`  | Both                       | When a stack `destroy` fails.     |
| `refresh_succeeded`    | `stack_update`  | Both                       | When a stack `refresh` succeeds.  |
| `refresh failed`       | `stack_update`  | Both                       | When a stack `refresh` fails.     |
| `deployment_queued`    | `deployment`    | Both                       | When a deployment is queued.      |
| `deployment_started`   | `deployment`    | Both                       | When a deployment starts running. |
| `deployment_succeeded` | `deployment`    | Both                       | When a deployment succeeds.       |
| `deployment_failed`    | `deployment`    | Both                       | When a deployment fails.          |

## Wrapping up

> We hope you and your team can streamline deployment related notifications to where you spend your time. Our webhooks improvements enable ChatOps workflows and more visibility into your infrastructure. As always, please let us know if you have feedback on the feature by opening an issue in the Pulumi Cloud requests repository or if you have other features you would like to see in Pulumi Cloud.

Reviewer 2: Should we link to the `Pulumi Cloud requests repository` here?

> We hope you and your team can streamline deployment related notifications to where you spend your time. Our webhooks improvements enable ChatOps workflows and more visibility into your infrastructure. As always, please let us know if you have feedback on the feature by opening an issue in the Pulumi Cloud requests repository or if you have other features you would like to see in Pulumi Cloud.

Reviewer 5: Same "streamline" comment as earlier, and also feels a reiteration of the opening line. How about something like this maybe?

```suggestion
These new features are all about getting you the information you need wherever you are, and we can't wait to see how you use them. Our webhooks improvements enable ChatOps workflows and more visibility into your infrastructure. As always, please let us know if you have feedback on the feature by opening an issue in the Pulumi Cloud requests repository or if you have other features you would like to see in Pulumi Cloud.
```
